src/mpid/ch4/src/ch4r_callbacks.c

```diff
@@ -281,20 +280,13 @@ int MPIDIG_send_target_msg_cb(int handler_id, void *am_hdr, void *data, MPI_Aint
          * set to the recv side datatype and count when it is matched */
         MPIDIG_REQUEST(rreq, datatype) = MPI_BYTE;
         MPIDIG_REQUEST(rreq, count) = hdr->data_sz;
-        if (hdr->data_sz) {
-            if (in_data_sz) {
-                /* If there is inline data, we allocate unexpected buffer */
-                MPIR_Assert(in_data_sz <= MPIR_CVAR_CH4_AM_PACK_BUFFER_SIZE);
-                mpi_errno =
-                    MPIDU_genq_private_pool_alloc_cell(MPIDI_global.unexp_pack_buf_pool, &pack_buf);
-                MPIR_Assert(pack_buf);
-                MPIDIG_REQUEST(rreq, buffer) = pack_buf;
-            } else {
-                /* if the SEND expect to contain data, but there is no inline data, mark the recv
-                 * request as not ready. */
-                MPIDIG_REQUEST(rreq, buffer) = NULL;
-                MPIDIG_REQUEST(rreq, recv_ready) = false;
-            }
+        if (in_data_sz) {
```

Comment:
I understand the logic is not new but it is never too late to add some comment --
We are using `in_data_sz` and `hdr->data_sz` together (which is implicit) to differentiate the *3*(or 4) modes here: 1. header only (zero data or RTS header), 2. Eager protocol and Eager transport mode, 3. Eager protocol and RNDV transport mode.  Ideally, the code structure should make these 3 (or 4) branches clear. The current code interweaves them, which is the root reason why this code is so difficult to read. We can refactor the code separately. Can you add some comments here explaining the 3 (or 4) distinct scenarios?